lib/Geo/ShapeContainer.cpp

```diff
@@ -571,27 +589,60 @@ bool ShapeContainer::intersects(S2Polyline const* other) const {
       return ll->Intersects(other);
     }
     case ShapeContainer::Type::S2_LATLNGRECT: {
-      S2LatLngRect const* rect = static_cast<S2LatLngRect const*>(_data);
-      for (int k = 0; k < other->num_vertices(); k++) {
-        if (rect->Contains(other->vertex(k))) {
-          return true;
-        }
-      }
-      return false;
+      auto rectPoly = ::latLngRectToPolygon(static_cast<S2LatLngRect const*>(_data));
+      auto cuts = rectPoly->IntersectWithPolyline(*other);
+      return !cuts.empty();
     }
     case ShapeContainer::Type::S2_POLYGON: {
       S2Polygon const* poly = static_cast<S2Polygon const*>(_data);
       auto cuts = poly->IntersectWithPolyline(*other);
       return !cuts.empty();
     }
     case ShapeContainer::Type::S2_MULTIPOINT:
-    case ShapeContainer::Type::S2_MULTIPOLYLINE:
+    case ShapeContainer::Type::S2_MULTIPOLYLINE: {
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_NOT_IMPLEMENTED,
+          "The case GEO_INTERSECTS(<multipoint or multipolyline>, <polyline>) is not yet implemented.");
+    }
     case ShapeContainer::Type::EMPTY:
       TRI_ASSERT(false);
   }
   return false;
 }
 
+namespace {
+bool intersectRectPolygon(S2LatLngRect const* rect, S2Polygon const* poly) {
+  if (rect->is_full()) {
+    return true;  // rectangle spans entire sphere
+  } else if (rect->is_point()) {
+    return poly->Contains(rect->lo().ToPoint());  // easy case
+  } else if (!rect->Intersects(poly->GetRectBound())) {
+    return false;  // cheap rejection
+  }
+  // Construct polygon from rect:
+  std::vector<S2Point> v;
+  v.reserve(5);
+  v.emplace_back(Coordinate::fromLatLng(rect->GetVertex(0)).toPoint());
+  v.emplace_back(Coordinate::fromLatLng(rect->GetVertex(1)).toPoint());
+  v.emplace_back(Coordinate::fromLatLng(rect->GetVertex(2)).toPoint());
+  v.emplace_back(Coordinate::fromLatLng(rect->GetVertex(3)).toPoint());
+  v.emplace_back(Coordinate::fromLatLng(rect->GetVertex(0)).toPoint());
+  std::unique_ptr<S2Loop> loop;
+  loop = std::make_unique<S2Loop>(std::move(v), S2Debug::DISABLE);
+  std::unique_ptr<S2Polygon> rectPoly;
+  rectPoly = std::make_unique<S2Polygon>(std::move(loop), S2Debug::DISABLE);
```

Comment:
that makes absolutely no sense to allocate `rectPoly` on a heap

Reply:
Right. Not that it would matter much, since the computation which is following is orders of magnitude more expensive than the memory allocation, but I fixed it.